app/src/main/java/de/westnordost/streetcomplete/measure/MeasureActivity.kt

```diff
@@ -0,0 +1,418 @@
+            anchor.detach()
+        }
+
+        try {
+            val newAnchor = hitResult.createAnchor()
+            val cursorNode = getCursorNode()
+            cursorNode.anchor = newAnchor
+
+            if (measureState == MeasureState.MEASURING) {
+                secondNode?.anchor = newAnchor
+            }
+            /* update distance should always be called because the world position could be adjusted
+             *  for already existing anchors */
+            updateDistance()
+        } catch (e: Exception) {
+            Log.e("MeasureActivity", "Error", e)
+        }
+    }
+
+    private fun updateDistance() {
+        val pos1 = firstNode?.worldPosition
+        val pos2 = secondNode?.worldPosition
+        val up = firstNode?.up
+        val hasMeasurement = pos1 != null && pos2 != null && up != null
+
+        binding.measurementSpeechBubble.isInvisible = !hasMeasurement
+        if (!hasMeasurement) return
+
+        val difference = Vector3.subtract(pos1, pos2)
+        distance = difference.length()
+
+        binding.measurementTextView.text = displayUnit.format(distance.toDouble())
+
+        val line = getLineNode()
+        line.worldPosition = Vector3.add(pos1, pos2).scaled(.5f)
+        line.worldRotation = Quaternion.lookRotation(difference, up)
+        line.localScale = Vector3(1f, 1f, distance)
+    }
+
+    private fun getCursorNode(): AnchorNode {
+        var node = cursorNode
+        if (node == null)  {
+            node = AnchorNode().apply {
+                renderable = cursorRenderable
+                setParent(arSceneView!!.scene)
+            }
+            cursorNode = node
+        }
+        return node
+    }
+
+    private fun getLineNode(): Node {
+        var node = lineNode
+        if (node == null)  {
+            node = Node().apply {
+                renderable = lineRenderable
+                setParent(arSceneView!!.scene)
+            }
+            lineNode = node
+        }
+        return node
+    }
+
+
+    /* ----------------------------------- Permission request ----------------------------------- */
+
+    /** Show dialog that explains why the camera permission is necessary. Returns whether the user
+     *  acknowledged the rationale. */
+    private suspend fun askUserToAcknowledgeCameraPermissionRationale(): Boolean =
+        suspendCancellableCoroutine { cont ->
+            val dlg = AlertDialog.Builder(this)
+                .setTitle(R.string.no_camera_permission_warning_title)
+                .setMessage(R.string.no_camera_permission_warning)
+                .setPositiveButton(android.R.string.ok) { _, _ -> cont.resume(true) }
+                .setNegativeButton(android.R.string.cancel) { _, _ -> cont.resume(false) }
+                .setOnCancelListener { cont.resume(false) }
+                .create()
+            cont.invokeOnCancellation { dlg.cancel() }
+            dlg.show()
+        }
```

Comment:
Why not more speaking variable names?

```suggestion
        suspendCancellableCoroutine { coroutine ->
            val dialog = AlertDialog.Builder(this)
                .setTitle(R.string.no_camera_permission_warning_title)
                .setMessage(R.string.no_camera_permission_warning)
                .setPositiveButton(android.R.string.ok) { _, _ -> coroutine.resume(true) }
                .setNegativeButton(android.R.string.cancel) { _, _ -> coroutine.resume(false) }
                .setOnCancelListener { coroutine.resume(false) }
                .create()
            coroutine.invokeOnCancellation { dialog.cancel() }
            dialog.show()
        }
```

Reply:
it's not a coroutine, it's a continuation. `cont` is used as an abbreviation of this throughout much of the codebase, same with `dlg` for dialog, so it is consistent.